crates/bevy_ecs/src/query/state.rs

```diff
@@ -74,6 +74,175 @@ pub struct QueryState<D: QueryData, F: QueryFilter = ()> {
     par_iter_span: Span,
 }
 
+/// View into a [`QueryState`] to avoid split-borrow issues when using [`QueryState`]
+pub struct QueryStateView<'me, D: QueryData, F: QueryFilter = ()> {
+    archetype_generation: &'me mut ArchetypeGeneration,
+    /// Metadata about the [`Table`](crate::storage::Table)s matched by this query.
+    matched_tables: &'me mut FixedBitSet,
+    /// Metadata about the [`Archetype`]s matched by this query.
+    matched_archetypes: &'me mut FixedBitSet,
+    /// [`FilteredAccess`] computed by combining the `D` and `F` access. Used to check which other queries
+    /// this query can run in parallel with.
+    component_access: &'me FilteredAccess<ComponentId>,
+    // NOTE: we maintain both a bitset and a vec because iterating the vec is faster
+    matched_storage_ids: &'me mut Vec<StorageId>,
+    fetch_state: &'me D::State,
+    filter_state: &'me F::State,
+}
+
+impl<'me, D: QueryData, F: QueryFilter> From<&'me mut QueryState<D, F>>
+    for QueryStateView<'me, D, F>
+{
+    fn from(value: &'me mut QueryState<D, F>) -> Self {
+        Self {
+            archetype_generation: &mut value.archetype_generation,
+            matched_tables: &mut value.matched_tables,
+            matched_archetypes: &mut value.matched_archetypes,
+            component_access: &value.component_access,
+            matched_storage_ids: &mut value.matched_storage_ids,
+            fetch_state: &value.fetch_state,
+            filter_state: &value.filter_state,
+        }
+    }
+}
+
+impl<'me, D: QueryData, F: QueryFilter> QueryStateView<'me, D, F> {
+    /// Updates the state's internal view of the `world`'s archetypes. If this is not called before querying data,
+    /// the results may not accurately reflect what is in the `world`.
+    ///
+    /// This is only required if a `manual` method (such as [`Self::get_manual`]) is being called, and it only needs to
+    /// be called if the `world` has been structurally mutated (i.e. added/removed a component or resource). Users using
+    /// non-`manual` methods such as [`QueryState::get`] do not need to call this as it will be automatically called for them.
+    ///
+    /// # Note
+    ///
+    /// This method only accesses world metadata.
+    ///
+    /// # Panics
+    ///
+    /// If `world` does not match the one used to call `QueryState::new` for this instance.
+    fn update_archetypes_unsafe_world_cell_with(
```

Comment:
Should this not be `pub` if `QueryState::get_manual` is pub?

Reply:
I don't think we want to make these pub, because the user could break the QueryState if they are not careful.

I introduce `QueryState::update_archetypes_unsafe_world_cell_with` and `QueryState::update_archetypes_with` mostly because we want the new logic to be used in both:
`QueryState::new` and `QueryState::new_with_access`

They both do something similar, but `new_with_access` also updates the provided access with each new archetype. Since that function is private, I think it's best to keep the other functions private.

I see your point that the docs mention public methods though, maybe the docs shouldn't mention these at all

Reply:
I was more worried if there was a way to produce invalid state with the `pub` methods mentioned that was irremediable because this is private. I don't think we should remove when it's intended to be used from the docs but maybe mention why it shouldn't be `pub`.